docs/usage/config.md
---
title: Configuring endpoints
---

# Configuring **GPTS** endpoints

**GPTS** uses declarative configuration which can be fetched or applied using `/config` endpoint.

!!! important "Supported formats"
    **GPTS** supports both JSON and YAML configuration formats, but uses JSON as default. In order to fetch/apply YAML configuration, the proper request header must be set:

    - `Accept: text/yaml` for `GET` requests
    - `Content-Type: text/yaml` for `POST` requests

## Configuration file structure

Declarative configuration used by **GPTS** is a map of URL paths (starting with `/`) and `Route` object, which consist of 7 fields:

| Field name | Description | Allowed values | Default value |
|-|-|-|-|
| `allowSubpaths` | Causes route to be used for all requests made to subpaths of defined URL path (e.g. if base path is `/abc`, then subpaths like `/abc/def`, `/abc/xyx` will be served using this configuration as well) | `true`, `false` | `false` |
| `default` | Default response (used when the specialized response for given method is not available) | `Response` object | `null` |
| `get` | Specialized response for GET request | `Response` object | `null` |
| `post` | Specialized response for POST request | `Response` object | `null` |
| `put` | Specialized response for PUT request | `Response` object | `null` |
| `patch` | Specialized response for PATCH request | `Response` object | `null` |
| `delete` | Specialized response for DELETE request | `Response` object | `null` |

`Response` object defines the details of expected response and consists of 4 fields:

| Field name | Description | Allowed values | Default value |
|-|-|-|-|
| `status` | HTTP status code | [correct HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) as integer | `200` |
| `contentType` | Value of "Content-Type" header; MIME type of returned content | valid MIME type | `text/plain` |
| `content` | Text content or `base64`-encoded binary content (in such case should start with `base64,`) | string | _empty string_ |
| `headers` | Dictionary of additional headers that should be sent in response | string to string map | `null` |

If the configuration is correct, **GPTS** will respond with `202 Accepted` status and no content. In such case, service will be reloaded with new configuration after approximately 2 seconds.

!!! warning "Missing configuration details"
    If neither default nor method-specific response is defined for given route, **GPTS** will respond with `503 Service Unavailable` status and no content.

!!! warning "Incorrect configuration"
    In case of trying to apply malformed configuration, **GPTS**'s `/config` endpoint will respond with following status codes and appropriate message:
    
    - `415 Unsupported Media Type` when sending content different from `application/json` or `text/yaml`
    - `400 Bad Request` if `JSON`/`YAML` content can't be parsed

!!! warning "Other errors"
    If any error would occur as the request is being processed (e.g. when decoding `base64`-encoded content), **GPTS** will respond with `500 Internal Server Error` status and no content.

!!! info "OpenAPI specification"
    For more formalized form of the description above, please look at [OpenAPI specification](https://editor.swagger.io/?url=https%3A%2F%2Fraw.githubusercontent.com%2FIcikowski%2FGPTS%2Fmaster%2Fopenapi.yml) of configuration endpoint.


## Fetching current configuration

In order to fetch current running configuration, `GET` request has to be made for `/config` endpoint.

=== "JSON"
    ```bash
    # Change ${GPTS} to proper GPTS service URL
    curl ${GPTS}/config
    ```

    ???- summary "Example command execution & output"
        ```bash
        curl -s http://localhost/config | jq
        ```
        ```json
        {
          "/hello": {
            "allowSubpaths": true,
            "default": {
              "status": 200,
              "contentType": "application/json",
              "content": "{\"message\":\"Hello World!\"}",
              "headers": {
                "X-SentBy": "GPTS - General Purpose Test Service"
              }
            }
          }
        }
        ```
=== "YAML"
    ```bash
    # Change ${GPTS} to proper GPTS service URL
    curl ${GPTS}/config -H "Accept: text/yaml"
    ```
    
    ???- summary "Example command execution & output"
        ```bash
        curl -s http://localhost/config -H "Accept: text/yaml" | yq eval -
        ```
        ```yaml
        /hello:
          allowSubpaths: true
          default:
            status: 200
            contentType: application/json
            content: '{"message":"Hello World!"}'
            headers:
              X-SentBy: GPTS - General Purpose Test Service
        ```

## Applying new configuration

In order to apply new configuration, `POST` request has to be made for `/config` endpoint with proper request body.

=== "JSON"
    ```bash
    curl -X POST http://localhost/config -H 'Content-Type: application/json' --data-binary @- <<EOF
    {
        // Content goes here...
    }
    EOF
    ```

    ???- summary "Example command execution & output"
        ```bash
        curl -X POST http://localhost/config -H 'Content-Type: application/json' -w "%{http_code}" --data-binary @- <<EOF
        {
          "/test": {
            "default": {
              "content": "this is some test content"
            }
          }
        }
        EOF
        ```
        ```
        202
        ```

=== "YAML"
    ```bash
    curl -X POST http://localhost/config -H 'Content-Type: text/yaml' --data-binary @- <<EOF
    # Content goes here
    EOF
    ```

    ???- summary "Example command execution & output"
        ```bash
        curl -X POST http://localhost/config -H 'Content-Type: text/yaml' -w "%{http_code}" --data-binary @- <<EOF
        /test:
          default:
            content: this is some test content
        EOF
        ```
        ```
        202
        ```


docs/usage/envvars.md
# Controlling **GPTS** setting with environment variables

There are four configurable **GPTS** settings which can be controlled using listed environment variables:

- `GPTS_SERVICE_PORT`,
- `GPTS_HEALTHCHECKS_PORT`,
- `GPTS_DEFAULT_CONFIG_ON_STARTUP`,
- `GPTS_LOG_LEVEL`,
- `GPTS_PRETTY_LOG`.

## `GPTS_SERVICE_PORT`

**Description**: Defines the port on which service will be exposed; defaults to `80` in most cases, but is set to `8080` in K8s deployments due to rootless mode[^1] limitations.

[^1]: **Rootless mode** is related to privileges level. **GPTS** images are built in the way that ensures usage of lowest possible privileges (user `app`, UID `1000`) in order to maintain proper security level. Due to lack of `root` privileges, the default service port (`80`) cannot be bind properly, therefore it was changed to `8080`.

**Allowed values**: valid port number

**Default value**: `80` (or `8080` in Helm charts)

## `GPTS_HEALTHCHECKS_PORT`

**Description**: Defines the port on which health endpoints will be exposed.

**Allowed values**: valid port number

**Default value**: `8081`

## `GPTS_DEFAULT_CONFIG_ON_STARTUP`

**Description**: Determines whether the default configuration (see below) should be loaded or not on application start.

**Allowed values**: `false` or `true`

**Default value**: `false`

???- example "Default configuration contents"
    === "JSON"
        ```json
        {
          "/hello": {
            "allowSubpaths": true,
            "default": {
              "status": 200,
              "contentType": "application/json",
              "content": "{\"message\":\"Hello World!\"}",
              "headers": {
                "X-SentBy": "GPTS - General Purpose Test Service"
              }
            }
          }
        }    
        ```
    === "YAML"
        ```yaml
        /hello:
          allowSubpaths: true
          default:
            status: 200
            contentType: application/json
            content: '{"message":"Hello World!"}' 
            headers:
              X-SentBy: GPTS - General Purpose Test Service
        ```

## `GPTS_LOG_LEVEL`

**Description**: Defines the desired log level in accordance to [`zerolog`'s log levels](https://github.com/rs/zerolog#leveled-logging).

**Allowed values**: `debug`, `info`, `warn`, `error`, `fatal`, `panic`, `trace`

**Default value**: `info`

## `GPTS_PRETTY_LOG`

**Description**: Enables/Disables pretty log format (may impact performance)

**Allowed values**: `false` or `true`

**Default value**: `false`


docs/changelog.md
---
title: Changelog
---

# Project changelog

## GPTS 0.7.3

- Changed liveness and readiness endpoints provider to `github.com/Icikowski/kubeprobes`
- Upgraded project's dependencies
- Updated documentation

## GPTS 0.7.2

- Updated chart definition
    - Added startup probe
    - Changed default values of `failureThreshold` and `periodSeconds` of all probes
    - Added `overrides` section to override name, fullname, image repository, image name & image tag
    - Added reference for GPTS' documentation for charts' values
    - Added metadata to `Chart.yaml`
- Added support for configurable configuration endpoint's address
- Upgraded project's dependencies
- Changed default handler's response structure
    - Headers' values are now returned as single comma-separated string instead of list
    - Added query values logging
- Updated OpenAPI specification
- Updated documentation

## GPTS 0.7.1

- Upgraded project's dependencies

## GPTS 0.7.0

- Changed base image to [Google's "distroless"](https://github.com/GoogleContainerTools/distroless)

## GPTS 0.6.6

- Fixed fetching configuration from environment variables

## GPTS 0.6.5

- Added support for command line flags (configuration may be passed either by CLI flags or environment variables, where flags are prioritized)

## GPTS 0.6.4

- Redesigned logging architecture
- Minor code improvements

## GPTS 0.6.3

- Added static binary compilation for standalone binaries and Docker image
- Enabled build information logging
- Added build information to binaries

## GPTS 0.6.2

- Renamed environment variables in Helm chart
- Renamed environment variables in Code
- Added log level selection to Helm chart
- Changed log levels of particular messages
- Added possibility to control desired log level
- Changed name of service readiness probe

## GPTS 0.6.1

- Fixed `YAML` vs `JSON` field names inconsistency

## GPTS 0.6.0

- Changed default ports in Helm chart
- Disabled loading default config on startup in Helm chart
- Changed Dockerfile to build rootless image

## GPTS 0.5.1

- Fixed AllowSubpaths option support
- Switched to `httptest.Server` from `http.Server` in tests

## GPTS 0.5.0

- Redesigned configuration structure
- Updated OpenAPI specification
- Updated Go version and project dependencies
- Extended tests coverage

## Early development versions

- Prepared Helm chart
- Prepared Dockerfile
- Added default configuration entry (`/hello` route)
- Added subpaths handling feature
- Introduced declarative configuration model
- Added default response for unconfigured routes
- Initial codebase of General Purpose Test Service


application/common/env.go
package common

import (
	"os"
	"strconv"
	"strings"
)

func getStringFromEnvironment(variableName string, fallback string) string {
	value, exists := os.LookupEnv(variableName)
	if !exists {
		return fallback
	}
	return value
}

func getBooleanFromEnvironment(variableName string, fallback bool) bool {
	value, exists := os.LookupEnv(variableName)
	if !exists {
		return fallback
	}
	value = strings.ToLower(value)
	return (value == "true" || value == "1" || value == "yes")
}

func getIntegerFromEnvironment(variableName string, fallback int) int {
	value, exists := os.LookupEnv(variableName)
	if !exists {
		return fallback
	}

	if parsedValue, err := strconv.Atoi(value); err != nil {
		return fallback
	} else {
		return parsedValue
	}
}

var (
	// ServicePort determines the port number on which service will be running (defaults to 80)
	ServicePort = getIntegerFromEnvironment("GPTS_SERVICE_PORT", 80)

	// HealthchecksPort determines the port number on which liveness & readiness endpoints will be running (defaults to 8081)
	HealthchecksPort = getIntegerFromEnvironment("GPTS_HEALTHCHECKS_PORT", 8081)

	// ConfigurationEndpoint determines the path of the configuration endpoint (defaults to /config)
	ConfigurationEndpoint = getStringFromEnvironment("GPTS_CONFIG_ENDPOINT", "/config")

	// DefaultConfigOnStartup determines if default config should be loaded when application starts (defaults to false)
	DefaultConfigOnStartup = getBooleanFromEnvironment("GPTS_DEFAULT_CONFIG_ON_STARTUP", false)

	// PrettyLog determines if pretty logging should be enabled (defaults to false)
	PrettyLog = getBooleanFromEnvironment("GPTS_PRETTY_LOG", false)

	// LogLevel determines the level of application log (defaults to "info")
	LogLevel = getStringFromEnvironment("GPTS_LOG_LEVEL", "info")
)
